### PR TITLE
chore(msgpack): fix schema printing brittleness

### PR DESCRIPTION
**Pull request by ludamad**
# Description

Since we moved to just one header in aztec, a brittleness with our previous scheme was encountered. This was reworked to use methods, which is also cleaner.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

**Comment by ludamad** [comment]
If accepted, will redo against the big wasi-20-sdk PR

File: cpp/src/barretenberg/plonk/proof_system/types/proof.hpp [hunk 1]
```diff
@@ -10,9 +10,10 @@ namespace proof_system::plonk {
 
 struct proof {
     std::vector<uint8_t> proof_data;
+    // For serialization, serialize as a buffer alias
     void msgpack_pack(auto& packer) const { packer.pack(proof_data); }
     void msgpack_unpack(auto object) { proof_data = (std::vector<uint8_t>)object; }
-
+    void msgpack_schema(auto& packer) const { packer.pack_alias("Proof", "bin32"); }
```
Comment (Maddiaa0) [inline cpp/src/barretenberg/plonk/proof_system/types/proof.hpp:16]: not familiar with this code, so forgive my ignorance, but why has this function gone to 
`void msgpack_schema` whilst the diff from `fr.hpp` goes to a `static constexpr char schema_name[] = "fr"`?

File: cpp/src/barretenberg/serialize/msgpack_impl/concepts.hpp [hunk 2]
```diff
@@ -4,12 +4,10 @@ struct DoNothing {
...
 namespace msgpack_concepts {
-template <typename T> concept HasMsgPack = requires(T t, DoNothing nop)
-{
-    t.msgpack(nop);
-};
-template <typename T> concept HasMsgPackPack = requires(T t, DoNothing nop)
-{
-    t.msgpack_pack(nop);
-};
+template <typename T>
```
Comment (Maddiaa0) [inline cpp/src/barretenberg/serialize/msgpack_impl/concepts.hpp:7]: nit: can a newline go between each of these concept definitions, this took my egg brain a while to parse

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp [hunk 3]
```diff
@@ -48,7 +53,46 @@ struct MsgpackSchemaPacker : msgpack::packer<msgpack::sbuffer> {
      * @tparam T the object's type.
      * @param obj the object.
      */
-    template <typename T> void pack_schema(const T& obj) { msgpack_schema_pack(*this, obj); }
+    template <typename T> void pack_schema(const T& obj) { msgpack_schema_pack_wrapper(*this, obj); }
+
+    // Recurse over any templated containers
+    // Outputs e.g. ['vector', ['sub-type']]
+    template <typename... Args> void pack_template_type(const std::string& schema_name)
```
Comment (Maddiaa0) [inline cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp:60]: Q from ignorance: Im not used to this pack syntax, is Args passed along outside of the parameters?

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp [hunk 4]
```diff
@@ -48,7 +53,46 @@ struct MsgpackSchemaPacker : msgpack::packer<msgpack::sbuffer> {
...
+    {
+        // We will pack a size 2 tuple
+        pack_array(2);
+        pack(schema_name);
+        pack_array(sizeof...(Args));
+
+        // Note: if this fails to compile, check first in list of template Arg's
+        // it may need a msgpack_schema_pack specialization (particularly if it doesn't define MSGPACK_FIELDS).
+        (msgpack_schema_pack_wrapper(*this, Args{}), ...); /* pack schemas of all template Args */
+    }
```
Comment (Maddiaa0) [inline cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp:70]: \n

**Review by Maddiaa0 — COMMENTED** [review]
I've had a look through and have questions, but at my current level of comprehension im in no position to review

File: cpp/src/barretenberg/plonk/proof_system/types/proof.hpp (hunk 1 above)
Comment (ludamad) [reply cpp/src/barretenberg/plonk/proof_system/types/proof.hpp:16]: It's a fairly meaningful refactor. Now, template types that are parameters have a schema_name, while other things have a schema method. It is meant to be cleaner

File: cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp (hunk 3 above)
Comment (ludamad) [reply cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp:60]: It is a type-only parameter

Comment (ludamad) [reply cpp/src/barretenberg/serialize/msgpack_impl/schema_impl.hpp:60]: Normally this wouldn't be useful, but as this is just printing schemas its all we need

**Review by dbanks12 — APPROVED** [review]
Hard to thoroughly follow all of these changes without spending more time, but LGTM. Just fix up some of @Maddiaa0's small requests and good to go.

**Comment by ludamad** [comment]
> Hard to thoroughly follow all of these changes without spending more time, but LGTM. Just fix up some of @Maddiaa0's small requests and good to go.

Thanks - if you have any feedback for me that could help that I'd be happy to try better next time. If it's just the nature of the thing then... thanks for the quick review :D

**Comment by dbanks12** [comment]
> Thanks - if you have any feedback for me that could help that I'd be happy to try better next time. If it's just the nature of the thing then... thanks for the quick review :D

Honestly I think it's the nature of the thing.... Maybe adding some more comments to the PR diff itself would help to give us hints at what some refactors are aiming to do. But yeah I think it's the nature of this msgpack beast.

**Pull request merged**